### PR TITLE
[New] Add dynamic entity layer

### PR DESCRIPTION
**Pull request by mhdostal**
## Description

This PR implements `Add dynamic entity layer` in `Maps` category.
URL to README: https://github.com/Esri/arcgis-maps-sdk-swift-samples/blob/mhd/AddDynamicEntityLayer/Shared/Samples/Add%20dynamic%20entity%20layer/README.md

## Linked Issue(s)

- `swift/issues/3734 `

## How To Test

- Tap "Disconnect/Connect" button and make sure connection status and observations are updated appropriately. When disconnected, observations will not be updated on the map.
- Open settings
- Toggle "Track lines" and "Previous observations" and ensure map updates appropriately
- Change "Observations per track" and ensure map updates to match
- Tap "Purge all observations" and ensure all observations are removed.

## To Discuss

Should the Model and/or Settings view go in separate files?
Doe the UI meet expectations and match existing samples?

### CODE REVIEW

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:11]:
```suggestion
A stream service is a type of service provided by ArcGIS Velocity and GeoEvent Server that allows clients to receive a stream of data observations via a web socket. ArcGIS Maps SDK allows you to connect to a stream service and manage the information as dynamic entities and display them in a dynamic entity layer. Displaying information from feeds such as a stream service is important in applications like dashboards where users need to visualize and track updates of real-world objects in real-time.
```

`common-samples/pull/3829`

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:21]:
```suggestion
1. Create an `ArcGIStreamService` using a stream service URL.
```
Most of the time if a design is not created by a Swift team member, it could contain phrasing that is specific to other platforms (in this case, .NET). Make sure to read through the whole README and make adjustments.

p.s., When migrate from an existing example, compare the old iOS README and the design - sometimes there are minor discrepancies between them. Pick the content that makes the most sense.

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:22]: I don't see the `DynamicEntityFilter` type in the API. Can you double check (with .NET, presumably Kerry) if this is the correct step, or we should use `ArcGISStreamServiceFilter` instead?

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:23]:
```suggestion
3. Set the `maximumDuration` property of the stream service's `purgeOptions` to limit the amount of data managed by the application.
```

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:25]:
```suggestion
5. Update values in the layer's `trackDisplayProperties` to customize the layer's appearance.
```

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:33]: As above, please check if this type is used.

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:35]: It seems this type is nested in Swift API, comparing to .NET. Change it to the correct type and reorder the list both here and in metadata.
```
* DynamicEntityDataSource.PurgeOptions
```

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/README.md:44]: Please discuss with realtime questions folks to get the link that should go here.

**Review by yo1995 — COMMENTED** [review]
First pass comments on the README.

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (mhdostal) [reply Shared/Samples/Add dynamic entity layer/README.md:21]: Yup should be `ArcGISStreamServiceFilter`.

**Comment by mhdostal** [comment]
I'm just waiting for feedback from the Milano team regarding the guide doc link.

**Comment by mhdostal** [comment]
The [word](https://esri-runtime.slack.com/archives/CBWAM7KME/p1679344960097719?thread_ts=1679340862.515769&cid=CBWAM7KME) from Milano is that the guid doc is currently being worked on.

@yo1995 I can remove that section of the README or leave it as is and updated it when the guide doc is ready.

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [reply Shared/Samples/Add dynamic entity layer/README.md:43]: Thanks. Let's remove the Additional information section from README. Once it is updated in the design I can add it back here.

**Review by vquach2404 — COMMENTED** [review]
Looks good so far! Seems like there's a glitch when pressing the info button

https://user-images.githubusercontent.com/52215269/226776001-cf78abc7-5950-4cfe-8deb-c4def8154158.mp4

**Comment by yo1995** [comment]
> Looks good so far! Seems like there's a glitch when pressing the info button

This is an issue with the sheet extension. #80

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift:38]:
```suggestion
```
Looks better to me this way.

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift:25]: If I turn "Previous observations" toggle on, it seems the track lines setting is overridden in the graphics, only shows the history blue dots, not the line. I'm assuming it is expected behavior.

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift:28]:
```suggestion
```


While it is green and inconsistent without this line, I tend to leave it out.

In the past we made a similar [change](https://github.com/Esri/arcgis-runtime-samples-ios/blob/main/arcgis-ios-sdk-samples/AppDelegate.swift#L112) in iOS Samples but now I'd like to follow Apple's default.

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:26]:
```suggestion
        let streamService: ArcGISStreamService = {
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:38]:
```suggestion
        let dynamicEntityLayer: DynamicEntityLayer
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:62]:
```suggestion
        /// The maximum observations range.
        /// Used by slider, which requires `CGFloat` values.
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:69]:
```suggestion
            // Creates the dynamic entity layer.
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:72]:
```suggestion
            // Initializes properties from the dynamic entity layer and stream service.
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:88]:
```suggestion
            // Adds the dynamic entity layer to the map's operational layers.
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:101]:
```suggestion
    /// A user-friendly string for `ConnectionStatus`.
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:86]: While it may be convenient to kick off the connection status "observation" in the initializer, another way we usually use is to add it to the `task` modifier on the main view suggested by @philium . A few benefits

- The task will have the same lifecycle as the main view, which is the equivalent to that of the sample
- No need to dispatch onto main queue when it is UI related
- No need to explicitly manage the `Task` in the model. When you create a task block it returns a handle that you can use to cancel or get the result from it. 

To my understanding, an unstructured task will have the same lifetime as its scope at creation, which is that of `Model`. It is not a problem in this sample, but IMO the less unstructured tasks, the less risk for a memory leak.

```swift
.task {
    // This will update `connectionStatus` when the stream service
    // connection status changes.
    for await status in model.streamService.$connectionStatus {
        model.connectionStatus = status.description
    }
}
```

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.swift:27]: I always love to refer to [xkcd 2170](https://xkcd.com/2170/) when I see sth like this 😆 

Coordinate Precision

No need to change, but we'll try our best to eliminate these minuscule particles next time when we review a sample design.

File: Shared/Samples/Add dynamic entity layer/README.md
Comment (yo1995) [reply Shared/Samples/Add dynamic entity layer/README.md:21]: Then it can be
```suggestion
2. Set an `ArcGISStreamServiceFilter` on the stream service to limit the amount of data coming from the server.
```

File: Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png:1]: Please update the screenshot if any UI related suggestion is applied above.

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift
Comment (yo1995) [inline Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.Model.swift:75]: No need to change, but I feel there is a risk if the server changes its default config `maximumObservations` beyond 1 ~ 16. Slider will show an OOB value.

File: Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png
Comment (yo1995) [reply Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png:1]: Thanks Vivian for noticing this. 9:41 status bar is highly recommended, but because a few Swift team folks don't readily have the [pre-bug](https://www.jessesquires.com/blog/2022/12/14/simctrl-status_bar-broken/) iOS 15/16.0 simulator, we might just let them grab whatever screenshot they can get… 🥲

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.swift
Comment (mhdostal) [reply Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.swift:27]: I love that one....

File: Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png
Comment (mhdostal) [reply Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png:1]: I still have some iOS 15.5 simulators installed, but running the script doesn't change the time/status bar. :-(

File: Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift
Comment (mhdostal) [reply Shared/Samples/Add dynamic entity layer/AddDynamicEntityLayerView.SettingsView.swift]: I'm not seeing that.  I think if you zoom in on a track you can see the lines.  When zoomed out, the lines are obscured by the dots.

File: Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png
Comment (mhdostal) [reply Shared/Samples/Add dynamic entity layer/add-dynamic-entity-layer-settings.png:1]: Done.

**Review by yo1995 — COMMENTED** [review]
Changes look good. Once the screenshots are replaced I'll approve.

**Comment by mhdostal** [comment]
> Changes look good. Once the screenshots are replaced I'll approve.

@yo1995 Done.

**Review by yo1995 — APPROVED** [review]

**Review by vquach2404 — APPROVED** [review]

**Comment by mhdostal** [comment]
@yo1995 @vquach2404 Thank you both!!

**Pull request merged**
